### PR TITLE
Support delaying attestation processing

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
Introduces a new `AttestationManager` which is roughly equivalent to the `BlockPropagationManager` but for attestations.  It can queue attestations that we're not yet ready to process until later - either based on the slot they are due to be processed in or blocks they depend on.

`PendingBlocks` and `FutureBlocks` have been adapted so they can be reused with attestations.

### CODE REVIEW

File: ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/ForkChoiceUtil.java [hunk 1]
```diff
@@ -417,65 +407,79 @@ private static boolean blockDescendsFromLatestFinalizedBlock(
...
-            <= 0,
-        "on_attestation: Attestations must not be for blocks in the future. If not, the attestation should not be considered");
+    if (get_current_slot(store).compareTo(target.getEpochSlot()) < 0) {
+      return AttestationProcessingResult.FAILED_FUTURE_EPOCH;
+    }
+
+    if (!store.getBlockRoots().contains(attestation.getData().getBeacon_block_root())) {
+      // Attestations must be for a known block. If block is unknown, delay consideration until the
+      // block is found
+      return AttestationProcessingResult.FAILED_UNKNOWN_TARGET;
```
Comment (mbaxter) [inline ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/ForkChoiceUtil.java]: Should this be renamed:
```suggestion
      return AttestationProcessingResult.FAILED_UNKNOWN_BLOCK;
```
?

File: sync/src/main/java/tech/pegasys/artemis/sync/AttestationManager.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        futureAttestations);
+  }
+
+  @Subscribe
+  @SuppressWarnings("unused")
+  private void onGossipedAttestation(final Attestation attestation) {
+    processAttestation(attestation);
+    // TODO: Should we post this if the attestation was invalid?
+    // What if it was delayed?
+    this.eventBus.post(new ProcessedAggregateEvent(attestation));
```
Comment (mbaxter) [inline sync/src/main/java/tech/pegasys/artemis/sync/AttestationManager.java]:
```suggestion
    this.eventBus.post(new ProcessedAttestationEvent(attestation));
```

File: sync/src/main/java/tech/pegasys/artemis/sync/AttestationManager.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        new ForkChoiceAttestationProcessor(storageClient, new StateTransition(false)),
+        pendingAttestations,
+        futureAttestations);
+  }
+
+  @Subscribe
+  @SuppressWarnings("unused")
+  private void onGossipedAttestation(final Attestation attestation) {
+    processAttestation(attestation);
+    // TODO: Should we post this if the attestation was invalid?
```
Comment (mbaxter) [inline sync/src/main/java/tech/pegasys/artemis/sync/AttestationManager.java:73]: I'd think we only want to post records if they're valid - did you mean to keep these TODO's in here?

File: sync/src/test/java/tech/pegasys/artemis/sync/AttestationManagerTest.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+import tech.pegasys.artemis.datastructures.operations.AttestationData;
+import tech.pegasys.artemis.datastructures.state.Checkpoint;
+import tech.pegasys.artemis.datastructures.util.DataStructureUtil;
+import tech.pegasys.artemis.statetransition.attestation.ForkChoiceAttestationProcessor;
+import tech.pegasys.artemis.statetransition.events.BlockImportedEvent;
+import tech.pegasys.artemis.storage.events.SlotEvent;
+import tech.pegasys.artemis.util.SSZTypes.Bitlist;
+import tech.pegasys.artemis.util.bls.BLSSignature;
+
+class AttestationManagerTest {
```
Comment (mbaxter) [inline sync/src/test/java/tech/pegasys/artemis/sync/AttestationManagerTest.java:49]: Should we add a couple of tests to make sure that `ProcessedAggregateEvent` and `ProcessedAttestationEvent` events are being posted when the corresponding aggregation / attestation is successfully processed?

**Review by mbaxter — APPROVED** [review]
LGTM except I think one of the events we're posting is the wrong type.

File: sync/src/main/java/tech/pegasys/artemis/sync/AttestationManager.java (hunk 2 above)
Comment (ajsutton) [reply sync/src/main/java/tech/pegasys/artemis/sync/AttestationManager.java]: Good spot.

File: ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/ForkChoiceUtil.java (hunk 1 above)
Comment (ajsutton) [reply ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/ForkChoiceUtil.java]: Yeah makes sense.

File: sync/src/main/java/tech/pegasys/artemis/sync/AttestationManager.java (hunk 3 above)
Comment (ajsutton) [reply sync/src/main/java/tech/pegasys/artemis/sync/AttestationManager.java:73]: The behaviour hasn't changed from what it was previously, but I added the TODO's because it seemed extremely weird that we'd always consider it processed.  I lean towards only posting these events once we actually have processed the attestations which may be delayed.  Will follow up in another PR. https://pegasys1.atlassian.net/browse/BC-243 created for good measure.

File: sync/src/test/java/tech/pegasys/artemis/sync/AttestationManagerTest.java (hunk 4 above)
Comment (ajsutton) [reply sync/src/test/java/tech/pegasys/artemis/sync/AttestationManagerTest.java:49]: Added basic tests and will flesh out the corner cases in the next PR which fixes them to do the right thing. :)
